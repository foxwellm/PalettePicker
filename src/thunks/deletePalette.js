import { setLoading, setError, removePalette } from '../actions';
import { fetchData } from '../utils/api';

export const deletePalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/palettes/${id}`, 'DELETE');
      dispatch(removePalette(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}