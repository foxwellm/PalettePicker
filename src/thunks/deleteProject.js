import { setLoading, setError, removeProject } from '../actions';
import { fetchData } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/projects/${id}`, 'DELETE');
      dispatch(removeProject(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}