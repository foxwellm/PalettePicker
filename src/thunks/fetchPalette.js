import { setLoading, setError, setPalettes } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const palette = await fetchData(`/api/v1/palettes/${id}`, 'GET');
      dispatch(setPalettes([palette]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};