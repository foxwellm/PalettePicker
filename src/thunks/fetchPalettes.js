import { setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalettes = (allProjects) => {
  return (dispatch) => {
    const promisedPalettes = allProjects.map(async palette => {
      try {
        let allPalettes = await fetchData(`/api/v1/projects/${palette.id}/palettes`, 'GET');
        allPalettes.sort((a, b) => {
          if (a.updated_at > b.updated_at) return -1;
          if (b.updated_at < a.updated_at) return 1;
          return 0;
        });
        return allPalettes
      } catch (error) {
        dispatch(setError(error.message));
      }
    })
    return Promise.all(promisedPalettes);
  }
}