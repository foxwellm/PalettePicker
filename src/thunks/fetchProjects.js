import { setLoading, setError, setProjects } from '../actions';
