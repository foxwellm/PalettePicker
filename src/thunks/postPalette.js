import { setLoading, setError, addPalette } from '../actions';
import { fetchData } from '../utils/api';

export const postPalette = (palette) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const newPalette = await fetchData('/api/v1/palettes', 'POST', palette);
      dispatch(addPalette(newPalette[0]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false))
  }
}