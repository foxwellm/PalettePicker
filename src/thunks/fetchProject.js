import { setLoading, setError, setProjects, setPalettes } from '../actions';
import { fetchData } from '../utils/api';
import { fetchPalettes } from './fetchPalettes';

export const fetchProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const project = await fetchData(`/api/v1/projects/${id}`, 'GET');
      dispatch(setProjects([project]));
      const projectPalettes = await dispatch(fetchPalettes([project]));
      dispatch(setPalettes(projectPalettes));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};