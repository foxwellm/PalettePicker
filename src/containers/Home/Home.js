import React, { Component } from 'react';
import { connect } from 'react-redux';
import Palette from '../Palette/Palette';
import ControlForm from '../ControlForm/ControlForm';
import ProjectContainer from '../ProjectContainer/ProjectContainer';
import { fetchProjects } from '../../thunks/fetchProjects';
import PropTypes from 'prop-types';

export class Home extends Component {
  
  componentDidMount = () => {
    this.props.fetchProjects();
  };

  render() {
    const { searching } = this.props;
    return (
      <div>
        {!searching && <Palette />}
        {!searching && <ControlForm />}
        <ProjectContainer />
      </div>
    );
  };
};

export const mapStateToProps = (state) => ({
  projects: state.projects,
  searching: state.searching,
});

export const mapDispatchToProps = (dispatch) => ({
  fetchProjects: () => dispatch(fetchProjects()),
});

Home.propTypes = {
  projects: PropTypes.array,
  searching: PropTypes.bool,
  fetchProjects: PropTypes.func,
}

Home.defaultProps = {
  projects: [],
  searching: false,
}

export default connect(mapStateToProps, mapDispatchToProps)(Home);