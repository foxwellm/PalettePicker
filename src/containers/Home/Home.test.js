import React from 'react';
import { Home, mapStateToProps, mapDispatchToProps } from './Home';
import { shallow } from 'enzyme';
import { fetchProjects } from '../../thunks/fetchProjects';
jest.mock('../../thunks/fetchProjects');

describe('Home', () => {
  let wrapper;
  let mockFetchProjects;
  let mockProjects;
  let mockSearching;

  beforeEach(() => {
    mockFetchProjects = jest.fn();
    mockProjects = [{ id: 1, name: 'Project 1'}, { id: 2, name: 'Project 2'}];
    mockSearching = true;

    wrapper = shallow(
      <Home 
        fetchProjects={mockFetchProjects}
        projects={mockProjects}
        searching={mockSearching}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot', () => {
    mockSearching = false;
    wrapper = shallow(
      <Home
        fetchProjects={mockFetchProjects}
        projects={mockProjects}
        searching={mockSearching}
      />
    )
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should call fetchProjects', () => {
      wrapper.instance().componentDidMount();
      expect(mockFetchProjects).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        projects: mockProjects,
        searching: mockSearching,
        otherState: false,
      };
      const expected = {
        projects: mockProjects,
        searching: mockSearching,
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when fetchProjects is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchProjects();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchProjects();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

});