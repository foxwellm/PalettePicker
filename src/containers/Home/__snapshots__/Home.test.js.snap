// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match the correct snapshot 1`] = `
<div>
  <Connect(ProjectContainer) />
</div>
`;

exports[`Home should match the correct snapshot 2`] = `
<div>
  <Connect(Palette) />
  <Connect(ControlForm) />
  <Connect(ProjectContainer) />
</div>
`;
